// Copyright 2014 The Swarming Authors. All rights reserved.
// Use of this source code is governed by the Apache v2.0 license that can be
// found in the LICENSE file.

// Messages for Primary <-> Replica auth DB replication protocol.
// Used from both Primary side (i.e. auth_service) and Replica side (any service
// that uses auth component).

package components.auth.proto.replication;


////////////////////////////////////////////////////////////////////////////////
// Linking protocol, used to associate Replicas with Primary.
//
// Overall protocol flow:
// 1. Primary generates and serializes ServiceLinkTicket message (via
//    /auth_service/api/v1/services/<app id>/linking_url call).
// 2. ServiceLinkTicket is passed to Replica as GET parameter in /auth/link
//    endpoint accessible only to GAE-level administrators of Replica instance.
// 3. Replica sends ServiceLinkRequest to Primary via direct service <-> service
//    URLFetch HTTP POST to /auth_service/api/v1/internal/link_replica.
// 4. Primary registers the replica, and replies with ServiceLinkResponse.
// 5. Replica registers the primary.
//
// Implementations:
//  * Replica side: components/auth/ui/ui.py
//  * Primary side: services/auth_service/frontend/handlers.py


// Generated by Primary, passed to Replica to initiate linking process.
message ServiceLinkTicket {
  // GAE application ID of Primary that generated this ticket. Replica will send
  // ServiceLinkRequest to this service when it processes the ticket.
  required string primary_id = 1;
  // URL to the root page of a primary service, i.e. https://<...>.appspot.com.
  // Useful when testing on dev appserver and on non-default version.
  required string primary_url = 2;
  // Identity of a user that generated this ticket.
  required string generated_by = 3;
  // Opaque blob passed back to Primary in ServiceLinkRequest. Its exact
  // structure is an implementation detail of Primary. It contains app_id of
  // a replica this ticket is intended for, timestamp and HMAC tag.
  required bytes ticket = 4;
}


// Sent from Replica to Primary via direct serivce <-> service HTTP call,
// replicas app_id would be available via X-Appengine-Inbound-Appid header.
message ServiceLinkRequest {
  // Same ticket that was passed to Replica via ServiceLinkTicket.
  required bytes ticket = 1;
  // URL to use when making requests to Replica from Primary.
  required string replica_url = 2;
  // Identity of a user that accepted the ticket and initiated this request.
  required string initiated_by = 3;
}


// Primary's response to ServiceLinkRequest. Always returned with HTTP code 200.
message ServiceLinkResponse {
  // Status codes.
  enum Status {
    // The service is now linked and primary will be pushing updates to it.
    SUCCESS = 0;
    // Primary do not replies.
    TRANSPORT_ERROR = 1;
    // Linking ticket is invalid or expired.
    BAD_TICKET = 2;
    // Linking ticket was generated for another app, not the calling one.
    AUTH_ERROR = 3;
  }
  required Status status = 1;
}
